go/plugins/firebase/retriever.go

```diff
@@ -25,34 +26,81 @@ import (
+const firestoreCollection = "FIRESTORE_COLLECTION"
+
 const (
-	Vector64 VectorType = iota
+	Vector64 VectorType = iota // Vector64 is the only supported vector type
 )
 
-const provider = "firebase"
-
+// RetrieverOptions struct for retriever configuration
 type RetrieverOptions struct {
-	Name            string
-	Label           string
-	Client          *firestore.Client
-	Collection      string
-	Embedder        ai.Embedder
-	VectorField     string
-	MetadataFields  []string
-	ContentField    string
-	Limit           int
-	DistanceMeasure firestore.DistanceMeasure
-	VectorType      VectorType
+	Name            string                    // Name of the retriever
+	Label           string                    // Label for the retriever
+	Collection      string                    // Firestore collection name
+	Embedder        ai.Embedder               // Embedder instance for generating embeddings
+	VectorField     string                    // Field name for storing vectors
+	MetadataFields  []string                  // List of metadata fields to retrieve
+	ContentField    string                    // Field name for storing content
+	Limit           int                       // Limit on the number of results
+	DistanceMeasure firestore.DistanceMeasure // Distance measure for vector similarity
+	VectorType      VectorType                // Type of vector (e.g., Vector64)
+}
+
+// Convert a Firestore document snapshot to a Genkit Document object.
+func convertToDoc(docSnapshots []*firestore.DocumentSnapshot, contentField string, metadataFields []string) []*ai.Document {
+	var documents []*ai.Document // Prepare the documents to return in the response
+
+	for _, result := range docSnapshots {
+		data := result.Data() // Retrieve document data
+
+		// Ensure content field exists and is of type string
+		content, ok := data[contentField].(string)
+		if !ok {
+			fmt.Printf("Content field %s missing or not a string in document %s", contentField, result.Ref.ID)
+			continue
+		}
+
+		// Extract metadata fields
+		metadata := make(map[string]interface{})
+		for _, field := range metadataFields {
+			if value, ok := data[field]; ok {
+				metadata[field] = value
+			}
+		}
+
+		// Create a Genkit Document object
+		doc := ai.DocumentFromText(content, metadata)
+		documents = append(documents, doc)
+	}
+	return documents
+}
+
+// Convert a slice of float32 to a slice of float64
+func convertToFloat64(queryEmbedding []float32) []float64 {
+	queryEmbedding64 := make([]float64, len(queryEmbedding)) // Allocate memory for float64 slice
+	for i, val := range queryEmbedding {
+		queryEmbedding64[i] = float64(val) // Convert each element
+	}
+	return queryEmbedding64
 }
 
-func DefineFirestoreRetriever(g *genkit.Genkit, cfg RetrieverOptions) (ai.Retriever, error) {
+// DefineFirestoreRetriever defines a retriever for Firestore
+func DefineFirestoreRetriever(g *genkit.Genkit, cfg RetrieverOptions, client *firestore.Client) (ai.Retriever, error) {
```

Comment:
```suggestion
func defineFirestoreRetriever(g *genkit.Genkit, cfg RetrieverOptions, client *firestore.Client) (ai.Retriever, error) {
```

So that it's not confused with the user-facing function.